### ISSUE TITLE
[BUG] Threading Segmentation Faults

### ISSUE THREAD

**Issue by willGraham01** [brainglobe/cellfinder#372]
First encountered [here](https://github.com/brainglobe/cellfinder/pull/369#issuecomment-1923605577) after attempting a fix for #368 |

From what we can infer; it appears that some combination of our multiprocessing, CPU allocation/requesting, and Python's garbage collection process is resulting in segmentation faults during _teardown_.

This only occurs on Ubuntu machines running Python 3.10 specifically, all other OSes are fine, as are other Python versions. In particular though, Python 3.10 is what BrainGlobe users on the HPC system have access too - and @alessandrofelder has received similar bug reports from users.

## The symptoms

Identifiable symptoms seem to be: after running the detection algorithm, but before completing teardown, Python will hang due to a segmentation fault in the (memory) cleanup phase.

This is most easily replicated by running the `tox` command to execute the tests on a GitHub runner. Note that the segmentation fault message will appear some time _after_ the first `test_detection` test runs.

- The fix on https://github.com/brainglobe/cellfinder/pull/369 seems to be to explicitly invoke the garbage collector at the end of the detection algorithm.
- Further evidence from the SLURM job outputs received by @alessandrofelder also suggest that the detection algorithm runs to completion, but then hangs at teardown due to bad/missing memory references.

## Minimum failing example

Will be pasted here once a script that reliably reproduces the error is available.

## Possible causes

- Multiprocessing memory management
- Delegation of tasks to cores
- Use of the `cpu_count` method allocating CPUs, and claiming they are "free", before garbage collection can be run

## Ruled out causes

- We are NOT requesting more CPUs than we have available, so this is not the cause of the problem
- The detection algorithm itself manages to complete without error, the issue always occurs after it finishes.

**Comment by alessandrofelder** [comment]
(Assuming what I observe on our local cluster is the same bug) and running with the latest changes from #375, it looks like https://github.com/brainglobe/cellfinder/blob/68ccb12d128c49eba75c1c916c36d6f8aa7e7d60/cellfinder/core/detect/filters/volume/volume_filter.py#L147 is reached, but https://github.com/brainglobe/cellfinder/blob/68ccb12d128c49eba75c1c916c36d6f8aa7e7d60/cellfinder/core/detect/filters/volume/volume_filter.py#L198 is not...

This is part of the code that happens still in the multiprocessing context, but should[ just happen in the main process.](https://brainglobe.info/community/developers/repositories/cellfinder-core/index.html#id1)

**Comment by willGraham01** [comment]
I'm even more confused because I created [this branch](https://github.com/brainglobe/cellfinder/pull/376) to attempt to find a minimum failing example... and everything looks like it's fine :cry:

> This is part of the code that happens still in the multiprocessing context, but should[ just happen in the main process.](https://brainglobe.info/community/developers/repositories/cellfinder-core/index.html#id1)

As in, you can see that it's still running in a multiprocessing context? Is there any way to force Python to only run a function on one process, and we can try and enforce that explicitly here to see if the error goes away?

**Comment by alessandrofelder** [comment]
It will be running inside the multiprocessing context we get on https://github.com/brainglobe/cellfinder/blob/68ccb12d128c49eba75c1c916c36d6f8aa7e7d60/cellfinder/core/detect/detect.py#L162

> Is there any way to force Python to only run a function on one process, and we can try and enforce that explicitly here to see if the error goes away?

Seems tricky in our code specifically at first glance... would require some non-trivial refactoring? 

> . and everything looks like it's fine 😢

The Windows and the possibly no-`numba` run are timing out though??

**Comment by willGraham01** [comment]
> The Windows and the possibly no-numba run are timing out though??

I set a timeout of 15 mins which is twice the length of time for the ubuntu seg-fault to occur. Those tests on Windows/ NUMBA disabled are actually running fine, but they're usually just slower than the others so get cancelled due to the timeout.

**Comment by willGraham01** [comment]
Alright, I think we are looking at two distinct issues here, since I have fixed the pytest problem we saw in CI.

## Bad cache restore on CI in #369 

This one _is_ an isolated GH problem.

Looking back at the logs, the cache that was downloaded for the original tests on the Ubuntu 3.10 job was smaller in size than all others for that job (although this has now retroactively changed thanks to the re-run and shared tasks). After a cache deletion everything is OK: https://github.com/brainglobe/cellfinder/actions/runs/7743237973/job/21314877990, and we don't need the garbage collection fix as shown on https://github.com/brainglobe/cellfinder/pull/376

TLDR it looks like attempt 1 didn't fetch the entire cache for some reason (maybe faulty download, not sure why). Forcing the branch to fetch the cache again seems to have done the trick. Will try merging in #376's branch, which removes all the "un-necessary" fixes, and re-running CI again after purging the cache once more.

## `brainmapper` hanging

_This_ however is a genuine issue with our multiprocessing logic, which we still need to get to the bottom of.

**Comment by alessandrofelder** [comment]
Moved the "brainmapper_hanging" issue to #383 as I think I have diagnosed the cause 🤞
